### PR TITLE
Major Refactor - Improved architecture, service patterns, and more!

### PR DESCRIPTION
**Pull request by carlank**
This, uh. I rewrote large swaths of the app, moving it towards a more standard architecture (similar to a normal API server, but with discord commands instead of API requests). Encapsulated lots of `common` stuff into services, which act as the middle layer between "controllers" (command handlers, not implemented fully) and data-level stuff. Should help with extensibility a lot, and prevent the duplication of concerns that was starting to happen. There are various bug fixes in here as well. 

I personally think this is a lot more understandable, but it's definitely different from how it was. lmk what you think @rbrtbrnschn 

Edit: I would recommend checking this branch out and looking at it locally! The diff is a bit hard to parse.

### CODE REVIEW

**Comment by carlank** [comment]
Oh, I'm gonna work on this later tonight, I'm just gonna be afk for a few hours and wanted to get this out there.

**Comment by carlank** [comment]
This could likely be merged at this point. There is more that I'm going to do in this vein, but it works as-is!

PLEASE LOOK BEFORE MERGING, though! This is a structural change, and will break any other current work based on main!

Let me know if you have any questions or concerns!

**Review by rbrtbrnschn — APPROVED** [review]
